### PR TITLE
add comfast wr610n support

### PR DESCRIPTION
**Pull request by baidxi**
Signed-off-by: juno <cl@comfast.cn>

Thanks for your contribution to OpenWrt!

To help keep the codebase consistent and readable,
and to help people review your contribution,
we ask you to follow the rules you find in the wiki at this link
https://openwrt.org/submitting-patches

Please remove this message before posting the pull request.

### CODE REVIEW

**Comment by adschm** [comment]
Hi, this PR has formal issue (commit title/message). Please check https://openwrt.org/submitting-patches once again.

File: target/linux/ath79/dts/qca9531_comfast_cf-wr610n.dts [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+
+	aliases {
+		serial0 = &uart;
+		label-mac-device = &eth1;
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		pinctrl-names = "default";
```
Comment (adschm) [inline target/linux/ath79/dts/qca9531_comfast_cf-wr610n.dts:20]: Do you need this line?

File: target/linux/ath79/dts/qca9531_comfast_cf-wr610n.dts [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+
+		wlan {
+			label = "green:wlan";
+			gpios = <&gpio 0 GPIO_ACTIVE_LOW>;
+			linux,default-trigger = "phy0tpt";
+		};
```
Comment (adschm) [inline target/linux/ath79/dts/qca9531_comfast_cf-wr610n.dts:26]: Please add empty lines between nodes.

File: target/linux/ath79/dts/qca9531_comfast_cf-wr610n.dts [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+			label = "blue:usr";
+			gpios = <&gpio 1 GPIO_ACTIVE_LOW>;
+		};
+		wan {
+			label = "green:wan";
+			gpios = <&gpio 2 GPIO_ACTIVE_LOW>;
+		};
+
+		lan {
+			label = "gree:lan";
```
Comment (adschm) [inline target/linux/ath79/dts/qca9531_comfast_cf-wr610n.dts:37]: green

File: target/linux/ath79/dts/qca9531_comfast_cf-wr610n.dts [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+			gpios = <&gpio 3 GPIO_ACTIVE_LOW>;
+		};
+
+		wan1 {
+			label = "green:wan1";
```
Comment (adschm) [inline target/linux/ath79/dts/qca9531_comfast_cf-wr610n.dts:42]: What's that LED for?

File: target/linux/ath79/dts/qca9531_comfast_cf-wr610n.dts [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+		lan2 {
+			label = "green:lan2";
+			gpios = <&gpio 15 GPIO_ACTIVE_LOW>;
+		};
+		lan1 {
+			label = "green:lan1";
+			gpios = <&gpio 16 GPIO_ACTIVE_LOW>;
+		};
+		wifi {
+			label = "blue:wifi";
```
Comment (adschm) [inline target/linux/ath79/dts/qca9531_comfast_cf-wr610n.dts:63]: What's the difference between wlan and wifi?

File: target/linux/ath79/dts/qca9531_comfast_cf-wr610n.dts [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+
+		reset {
+			label = "reset";
+			linux,code = <KEY_RESTART>;
+			gpios = <&gpio 17 GPIO_ACTIVE_LOW>;
+			debounce-interval = <60>;
+		};
+	};
+
+	watchdog:watchdog {
```
Comment (adschm) [inline target/linux/ath79/dts/qca9531_comfast_cf-wr610n.dts:79]: Do you need the DT label here?

File: target/linux/ath79/dts/qca9531_comfast_cf-wr610n.dts [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+		};
+	};
+};
+
+&uart {
+	status = "okay";
+};
+
+&eth0 {
+	status = "okay";
```
Comment (adschm) [inline target/linux/ath79/dts/qca9531_comfast_cf-wr610n.dts:134]: Please add an empty line after status.

File: target/linux/ath79/dts/qca9531_comfast_cf-wr610n.dts [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+	mtd-mac-address = <&art 0x6>;
+	phy-handle = <&swphy0>;
+};
+
+&eth1 {
+	mtd-mac-address = <&art 0x0>;
```
Comment (adschm) [inline target/linux/ath79/dts/qca9531_comfast_cf-wr610n.dts:140]: Does the MAC address assignment match vendor firmware behavior?

File: target/linux/ath79/dts/qca9531_comfast_cf-wr610n.dts [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+	gmac-config {
+		device = <&gmac>;
+		switch-phy-swap = <0>;
+	};
+};
+
+&wmac {
+	status = "okay";
+
+	mtd-mac-address = <&art 1002>;
```
Comment (adschm) [inline target/linux/ath79/dts/qca9531_comfast_cf-wr610n.dts:151]: This line can be dropped, it is automatically extracted from caldata.

File: target/linux/ath79/generic/base-files/etc/board.d/02_network [hunk 10]
```diff
@@ -114,6 +114,7 @@ ath79_setup_interfaces()
 	tplink,tl-wdr3500-v1|\
 	tplink,tl-wr842n-v1|\
 	tplink,tl-wr842n-v3|\
+	comfast,cf-wr610n|\
```
Comment (adschm) [inline target/linux/ath79/generic/base-files/etc/board.d/02_network:117]: Please take care of alphabetic sorting.

File: target/linux/ath79/image/generic.mk [hunk 11]
```diff
@@ -529,6 +529,15 @@ define Device/comfast_cf-wr752ac-v1
 endef
 TARGET_DEVICES += comfast_cf-wr752ac-v1
 
+define Device/comfast_cf-wr610n
```
Comment (adschm) [inline target/linux/ath79/image/generic.mk:532]: Please take care of alphabetic sorting. No v1/v2?

File: target/linux/ath79/image/generic.mk [hunk 12]
```diff
@@ -529,6 +529,15 @@ define Device/comfast_cf-wr752ac-v1
...
+  SOC := qca9531
+  DEVICE_VENDOR := COMFAST
+  DEVICE_MODEL := CF-WR610N
+  DEVICE_PACKAGES := swconfig -uboot-envtools
```
Comment (adschm) [inline target/linux/ath79/image/generic.mk:536]: swconfig is added by default

**Comment by adschm** [comment]
No setup for lan/wan LEDs?

File: target/linux/ath79/dts/qca9531_comfast_cf-wr610n.dts (hunk 6 above)
Comment (baidxi) [reply target/linux/ath79/dts/qca9531_comfast_cf-wr610n.dts:79]: yes, this device has 706 hardware gpio watchdog.

File: target/linux/ath79/dts/qca9531_comfast_cf-wr610n.dts (hunk 8 above)
Comment (baidxi) [reply target/linux/ath79/dts/qca9531_comfast_cf-wr610n.dts:140]: yes, art partition offset 0 for lan mac, offset 6 for wan.

File: target/linux/ath79/generic/base-files/etc/board.d/02_network (hunk 10 above)
Comment (baidxi) [reply target/linux/ath79/generic/base-files/etc/board.d/02_network:117]: ok

File: target/linux/ath79/image/generic.mk (hunk 11 above)
Comment (baidxi) [reply target/linux/ath79/image/generic.mk:532]: ok, only this version.

> No setup for lan/wan LEDs?

yes, ALL PORT used default

File: target/linux/ath79/dts/qca9531_comfast_cf-wr610n.dts (hunk 6 above)
Comment (adschm) [reply target/linux/ath79/dts/qca9531_comfast_cf-wr610n.dts:79]: "DT label" = "watchdog:"

I do think this is not needed, and you just need `watchdog {` here.

File: target/linux/ath79/image/generic.mk (hunk 11 above)
Comment (adschm) [reply target/linux/ath79/image/generic.mk:532]: If the WAN/LAN LEDs work automatically without any specific setup, then you probably used the wrong gpio numbers.

**Review by baidxi — COMMENTED** [review]
finish change and tested.
led work OK.
network ok.
wifi ok.
status led ok

File: target/linux/ath79/dts/qca9531_comfast_cf-wr610n.dts (hunk 1 above)
Comment (baidxi) [reply target/linux/ath79/dts/qca9531_comfast_cf-wr610n.dts:23]: this device front have a three color led for wifi status, sys status and wan port status.

File: target/linux/ath79/generic/base-files/etc/board.d/01_leds [hunk 13]
```diff
@@ -140,6 +140,9 @@ comfast,cf-e560ac)
 	ucidef_set_led_switch "lan3" "LAN3" "blue:lan3" "switch0" "0x08"
 	ucidef_set_led_switch "lan4" "LAN4" "blue:lan4" "switch0" "0x10"
 	;;
+comfast,cf-wr610n)
```
Comment (adschm) [inline target/linux/ath79/generic/base-files/etc/board.d/01_leds:143]: Alphabetic sorting ...

File: target/linux/ath79/image/generic.mk (hunk 11 above)
Comment (adschm) [reply target/linux/ath79/image/generic.mk:532]: Still need to fix alphabetic sorting here.

**Comment by adschm** [comment]
Please do not set the execute bit on platform.sh and generic.mk.

When all changes are done, please squash your changes into a proper, single commit (git rebase -i ...) and force-push it to this PR (git push -f ...). Then you also still need to apply proper commit title, message etc.

**Review by baidxi — COMMENTED** [review]
all changes are done

**Comment by adschm** [comment]
> all changes are done

Err, no?

**Comment by baidxi** [comment]
yes，no err。发自我的华为手机-------- 原始邮件 --------发件人： Adrian Schmutzler <notifications@github.com>日期： 2020年10月31日周六 02:09收件人： openwrt/openwrt <openwrt@noreply.github.com>抄送： baidxi <wskityiago@163.com>, Author <author@noreply.github.com>主    题： Re: [openwrt/openwrt] add comfast wr610n support (#3515)

all changes are done

Err, no?

**Comment by baidxi** [comment]
> > all changes are done
> 
> Err, no?

yes validationed yet

**Review by baidxi — COMMENTED** [review]
all ok

**Review by baidxi — COMMENTED** [review]
modify done

**Comment by baidxi** [comment]
> > all changes are done
> 
> Err, no?

are you have some time to review my code ?

File: target/linux/ath79/dts/qca9531_comfast_cf-wr610n.dts [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+
+/ {
+	compatible = "comfast,cf-wr610n", "qca,qca9531";
+	model = "COMFAST CF-WR610N";
+
+	aliases {
+		serial0 = &uart;
+		led-boot = &wan;
```
Comment (adschm) [inline target/linux/ath79/dts/qca9531_comfast_cf-wr610n.dts:14]: I still don't like using the wan LED for boot indication.

File: target/linux/ath79/dts/qca9531_comfast_cf-wr610n.dts [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+	leds {
+		compatible = "gpio-leds";
+
+		wifi {
+			label = "green:wifi";
+			gpios = <&gpio 0 GPIO_ACTIVE_HIGH>;
+			linux,default-trigger = "phy0tpt";
+		};
+
+		failsafe:failsafe {
```
Comment (adschm) [inline target/linux/ath79/dts/qca9531_comfast_cf-wr610n.dts:29]: `led_failsafe: failsafe {`

File: target/linux/ath79/dts/qca9531_comfast_cf-wr610n.dts [hunk 16]
```diff
@@ -0,0 +1,125 @@
...
+			label = "red:failsafe";
+			gpios = <&gpio 2 GPIO_ACTIVE_HIGH>;
+		};
+
+		wan:wan {
```
Comment (adschm) [inline target/linux/ath79/dts/qca9531_comfast_cf-wr610n.dts:34]: `led_wan: wan {`

or just

`wan {`

if you remove this from led-boot ..

**Comment by adschm** [comment]
Would be nice to remove the unneeded global indent from the commit message.

**Comment by adschm** [comment]
Would you please provide a link to a data sheet or similar where the LEDs are described/listed?

And, completely different subject: commit title has a typo: suport->support

**Comment by baidxi** [comment]
> Would you please provide a link to a data sheet or similar where the LEDs are described/listed?
> 
> And, completely different subject: commit title has a typo: suport->support

this device have not other led indicate device status on boot and boot failed. 

so,  if removed,  on booting all led is off status, we don't know device boot status.

File: target/linux/ath79/dts/qca9531_comfast_cf-wr610n.dts (hunk 16 above)
Comment (baidxi) [reply target/linux/ath79/dts/qca9531_comfast_cf-wr610n.dts:34]: hi, can you review ?

**Comment by adschm** [comment]
> Would you please provide a link to a data sheet or similar where the LEDs are described/listed?

**Comment by baidxi** [comment]
> > Would you please provide a link to a data sheet or similar where the LEDs are described/listed?

like this.
[https://item.taobao.com/item.htm?id=547421824199&ali_refid=a3_420435_1006:1105317676:N:8VjXpceWiJC1YHHCgfpr8dMC9he8qZaB:01c9191802a5bbf1f7c447aad6133974&ali_trackid=1_01c9191802a5bbf1f7c447aad6133974&spm=a230r.1.1989828.13
](url)

**Comment by baidxi** [comment]
> > > Would you please provide a link to a data sheet or similar where the LEDs are described/listed?
> 
> like this.
> [https://item.taobao.com/item.htm?id=547421824199&ali_refid=a3_420435_1006:1105317676:N:8VjXpceWiJC1YHHCgfpr8dMC9he8qZaB:01c9191802a5bbf1f7c447aad6133974&ali_trackid=1_01c9191802a5bbf1f7c447aad6133974&spm=a230r.1.1989828.13 ](url)

rgb led list added.

**Comment by baidxi** [comment]
> > Would you please provide a link to a data sheet or similar where the LEDs are described/listed?

rgb led list added.

File: target/linux/ath79/dts/qca9531_comfast_cf-wr610n.dts [hunk 17]
```diff
@@ -0,0 +1,125 @@
...
+		gpios = <&gpio 13 GPIO_ACTIVE_LOW>;
+		hw_algo = "toggle";
+		hw_margin_ms = <1000>;
+	};
+};
+
+&spi {
+	status = "okay";
+
+	num-cs = <1>;
```
Comment (adschm) [inline target/linux/ath79/dts/qca9531_comfast_cf-wr610n.dts:62]: num-cs can probably be dropped.

**Comment by adschm** [comment]
There are still old comments not addressed ...

**Pull request closed without merging**